source/validations.ts

```diff
@@ -0,0 +1,109 @@
+// /source/validations.ts
+// The validation functions
+
+import type { Request } from 'express'
+import { ValidationError } from './types.js'
+
+/**
+ * Runs the given validation check, and conveys the error message to the user.
+ *
+ * @param shouldRun {boolean} - Whether or not the check should even run.
+ * @param check {Function} - The validation check to run.
+ * @param error {ValidationError} - The error to convey in case the validation check fails.
+ *
+ * @returns {void}
+ */
+const runCheck = (
+	shouldRun: boolean,
+	check: () => boolean,
+	error: ValidationError,
+): void => {
+	if (!shouldRun) return
+
+	// Run the check. If it fails, then return the error message to the user.
+	if (!check()) {
+		const formattedMessage = `express-rate-limit: ${error.code} ${error.message}`
+		console.warn(formattedMessage)
+	}
+}
+
+// The following regex matches IPv4 and IPv6, but not if they include a port number at the end.
+// https://www.regextester.com/104038
+const ipAddressRegex =
+	/((^\s*(((\d|[1-9]\d|1\d{2}|2[0-4]\d|25[0-5])\.){3}(\d|[1-9]\d|1\d{2}|2[0-4]\d|25[0-5]))\s*$)|(^\s*((([\dA-Fa-f]{1,4}:){7}([\dA-Fa-f]{1,4}|:))|(([\dA-Fa-f]{1,4}:){6}(:[\dA-Fa-f]{1,4}|((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3})|:))|(([\dA-Fa-f]{1,4}:){5}(((:[\dA-Fa-f]{1,4}){1,2})|:((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3})|:))|(([\dA-Fa-f]{1,4}:){4}(((:[\dA-Fa-f]{1,4}){1,3})|((:[\dA-Fa-f]{1,4})?:((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3}))|:))|(([\dA-Fa-f]{1,4}:){3}(((:[\dA-Fa-f]{1,4}){1,4})|((:[\dA-Fa-f]{1,4}){0,2}:((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3}))|:))|(([\dA-Fa-f]{1,4}:){2}(((:[\dA-Fa-f]{1,4}){1,5})|((:[\dA-Fa-f]{1,4}){0,3}:((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3}))|:))|(([\dA-Fa-f]{1,4}:)(((:[\dA-Fa-f]{1,4}){1,6})|((:[\dA-Fa-f]{1,4}){0,4}:((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3}))|:))|(:(((:[\dA-Fa-f]{1,4}){1,7})|((:[\dA-Fa-f]{1,4}){0,5}:((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3}))|:)))(%.+)?\s*$))/
+
+/**
+ * Checks whether the IP address is valid, and that it does not have a port
+ * number in it.
+ *
+ * The port from which a request comes can be changed simply by opening or
+ * closing a browser, or when using an Azure proxy, thus opening up avenues
+ * for bypassing the rate limit.
+ *
+ * See https://github.com/express-rate-limit/express-rate-limit/wiki/Troubleshooting-Proxy-Issues.
+ *
+ * @param ip {string | undefined} - The IP address provided by Express as request.ip.
+ *
+ * @returns {void}
+ */
+export const validateIp = (shouldRun: boolean, ip: string | undefined) => {
+	runCheck(
+		shouldRun,
+		() => ipAddressRegex.test(ip ?? 'no-ip'),
```

Comment:
When express-rate-limit is used correctly, `req.ip` shouldn't ever be `undefined`, so I don't think we should put that in our type signature. (I wanted the test to check for that, but `undefined as any` should be fine there.)

Then this should work without the `?? 'no-ip'`:

```
> var a
undefined
> /[0-9-]+/.test(a)
false
```
(`undefined` is automatically coerced to a string as needed in JS):
```
> /undefined/.test(a)
true
```

Reply:
Ideally, yes, `request.ip` would never be undefined. However, there might be a misconfiguration that causes the value to be undefined, and since the purpose of the check is to make sure there's a valid IP, I check for undefined values too.

Reply:
The `?? 'no-ip'` part is to please the linter, it doesn't like templates being possibly undefined.

Reply:
What if we made it two separate checks?  We could first run the original check to see if `req.ip` is even defined, and then if that passes, run this check to see if it's valid. They're distinct enough issues that having separate error messages probably makes sense.

I'm also thinking about setting up wiki pages (or perhaps a website) for each error with a bit more info, and then linking to it from the error message.

Reply:
Sure, that sounds good.

A website seems a bit overkill, I think a wiki page with all the error codes will do. We already have some info in the [Troubleshooting Proxy Issues](https://github.com/express-rate-limit/express-rate-limit/wiki/Troubleshooting-Proxy-Issues) section, we can use that :)

Reply:
I added the [Error Codes](https://github.com/express-rate-limit/express-rate-limit/wiki/Error-Codes) page to the wiki, and linked to it in the error messages.

Reply:
Thank you, that's amazing!

The only reason I was thinking about having a website would be to have shorter URLs; they could even redirect to the wiki. But, I think just putting the wiki link in there is fine for now.